src/utils/deduper.zig

```diff
@@ -0,0 +1,203 @@
+
+const Instant = sig.time.Instant;
+const Duration = sig.time.Duration;
+const AHashSeed = sig.utils.ahash.AHashSeed;
+const AHasher = sig.utils.ahash.AHasher;
+const ChaChaRng = sig.rand.ChaChaRng(20);
+const uintLessThanRust = sig.rand.weighted_shuffle.uintLessThanRust;
+
+/// Deduper which uses the AHash algorithm across K different seeds/hashers to find possible duplicates.
+pub fn Deduper(comptime n_hashers: usize, comptime T: type) type {
+    return struct {
+        num_bits: u64,
+        bits: std.ArrayList(AtomicU64),
+        state: [n_hashers]AHashSeed,
+        last_reset_instant: Instant,
+        masked_count: AtomicU64,
+
+        pub fn init(
+            allocator: std.mem.Allocator,
+            rand: std.rand.Random,
+            num_bits: u64,
+        ) !Deduper(n_hashers, T) {
+            const size: usize = @intCast((num_bits + 63) / 64);
+            var bits = try std.ArrayList(AtomicU64).initCapacity(allocator, size);
+            for (0..size) |_| bits.appendAssumeCapacity(AtomicU64.init(0));
+            var state: [n_hashers]AHashSeed = undefined;
+            for (0..n_hashers) |i| state[i] = AHashSeed.fromRng(rand);
+            return .{
+                .num_bits = num_bits,
+                .bits = bits,
+                .state = state,
+                .last_reset_instant = Instant.now(),
+                .masked_count = AtomicU64.init(0),
+            };
+        }
+
+        pub fn deinit(self: Deduper(n_hashers, T)) void {
+            self.bits.deinit();
+        }
+
+        /// Resets the deduper if the false positive rate is too high or the reset cycle has elapsed.
+        pub fn maybeReset(
+            self: *Deduper(n_hashers, T),
+            rand: std.rand.Random,
+            false_positive_rate: f64,
+            reset_cycle: Duration,
+        ) bool {
+            std.debug.assert(0.0 < false_positive_rate and false_positive_rate < 1.0);
+            const saturated = self.falsePositiveRate() >= false_positive_rate;
+            if (saturated or
+                self.last_reset_instant.elapsed().asNanos() >= reset_cycle.asNanos())
+            {
+                for (self.bits.items) |_bit| {
+                    var bit = _bit;
+                    bit = AtomicU64.init(0);
+                }
+                for (0..n_hashers) |i| self.state[i] = AHashSeed.fromRng(rand);
+                self.last_reset_instant = Instant.now();
+                self.masked_count = AtomicU64.init(0);
+            }
+            return saturated;
+        }
+
+        /// Returns true if the data is likely a duplicate.
+        pub fn dedup(self: *Deduper(n_hashers, T), data: *const T) bool {
+            var duplicate = true;
+            for (0..n_hashers) |i| {
+                var hasher = AHasher.fromSeed(self.state[i]);
+                hasher.hash(T, data);
+                const hash: u64 = hasher.finish() % self.num_bits;
+                const mask: u64 = @as(u64, 1) << @truncate(hash);
```

Comment:
The hash output is 64 bits and this only utilizes 6 of those bits. Then it loops around and hashes again to get another 6 bits. But you don't actually need to hash on every iteration since you still have more bits available. Why not use the remaining bits from the prior hash until all the bits are consumed? That way you could reduce the number of hash calculations and hashers by a factor of 10.